File: src/sdk/ts/src/intent-graph-storage.ts
import { readFile, rename, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { IntentGraph } from "./index.js";
import { resolveS3Endpoint, signS3Request } from "./sigv4.js";

/**
 * Host-owned persistence for an {@link IntentGraph} (ADR-0025). Core stays
 * bytes-in/bytes-out (ADR-0014) — these are thin adapters over
 * `IntentGraph.toJson()`/`fromJson()`/`rev` that live entirely in the SDK.
 *
 * Both implementations use `rev` for two things: **save-when-changed** (skip
 * the write if `rev` hasn't moved since the last save) and **stale-base
 * detection** (raise {@link StaleIntentGraphError} instead of clobbering a
 * concurrent writer — single-writer model, detect don't merge).
 */
export interface ExperimentalIntentGraphStorage {
  /** Load the stored graph, or `null` if nothing has been saved yet. */
  load(): Promise<IntentGraph | null>;
  /** Save `graph`, or skip if unchanged since the last save/load `rev`. */
  save(graph: IntentGraph): Promise<void>;
}

/** Another writer saved a newer graph since this storage object's last `load()`/`save()`. */
export class StaleIntentGraphError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StaleIntentGraphError";
  }
}

/** Options for {@link ExperimentalLocalFileIntentGraphStorage}. */
export interface ExperimentalLocalFileIntentGraphStorageOptions {
  /** Path to the JSON file. Parent directory must already exist. */
  readonly path: string;
}

/**
 * Local JSON file storage for an {@link IntentGraph} — the default backend.
 * Writes atomically (temp file + rename) so a crash mid-write cannot leave a
 * truncated file.
 */
export class ExperimentalLocalFileIntentGraphStorage implements ExperimentalIntentGraphStorage {
  private readonly path: string;
  private lastKnownRev: number | undefined;

  constructor(options: ExperimentalLocalFileIntentGraphStorageOptions) {
    this.path = options.path;
  }

  async load(): Promise<IntentGraph | null> {
    let text: string;
    try {
      text = await readFile(this.path, "utf8");
    } catch (error) {
      if (isNotFound(error)) {
        this.lastKnownRev = undefined;
        return null;
      }
      throw error;
    }
    const graph = IntentGraph.fromJson(text);
    this.lastKnownRev = graph.rev;
    return graph;
  }

  async save(graph: IntentGraph): Promise<void> {
    if (this.lastKnownRev === graph.rev) return;

    const diskRev = await this.readDiskRev();
    if (diskRev !== this.lastKnownRev) {
      throw new StaleIntentGraphError(
        `intent graph at ${this.path} changed since load() (on-disk rev ${diskRev ?? "none"}, ` +
          `expected ${this.lastKnownRev ?? "none"}); load() again and reapply your changes before saving`,
      );
    }

    const tmpPath = join(
      dirname(this.path),
      `.tmp-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}`,
    );
    await writeFile(tmpPath, graph.toJson(), "utf8");
    await rename(tmpPath, this.path);
    this.lastKnownRev = graph.rev;
  }

  private async readDiskRev(): Promise<number | undefined> {
    try {
      const text = await readFile(this.path, "utf8");
      return (JSON.parse(text) as { rev?: number }).rev ?? 0;
    } catch (error) {
      if (isNotFound(error)) return undefined;
      throw error;
    }
  }
}

function isNotFound(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && error.code === "ENOENT";
}

/**
 * Extract `<Code>`/`<Message>` from an S3 error response body (standard AWS
 * XML error shape) for a more useful failure message than a bare status code
 * — e.g. distinguishing `SignatureDoesNotMatch` from `AccessDenied` on a 403.
 * Returns `undefined` for a body with no `<Code>` (not an AWS-shaped error).
 */
function describeS3Error(body: string): string | undefined {
  const code = /<Code>([^<]*)<\/Code>/.exec(body)?.[1];
  if (!code) return undefined;
  const message = /<Message>([^<]*)<\/Message>/.exec(body)?.[1];
  return message ? `${code}: ${message}` : code;
}

/** One S3 REST request, as {@link S3Transport} sends it. */
export interface S3Request {
  /** `"GET"` (read the object) or `"PUT"` (write it). */
  readonly method: "GET" | "PUT";
  /** S3 bucket name. */
  readonly bucket: string;
  /** S3 object key. */
  readonly key: string;
  /** Lowercase header names, e.g. `if-match`, `if-none-match`. */
  readonly headers: Readonly<Record<string, string>>;
  /** Request body for a `PUT`; omitted for a `GET`. */
  readonly body?: string;
}

/** Response from an {@link S3Transport} call. */
export interface S3Response {
  /** HTTP status code, e.g. `200`, `404`, `412`. */
  readonly status: number;
  /** Lowercase header names. */
  readonly headers: Readonly<Record<string, string>>;
  /** Response body (the object contents for a successful `GET`). */
  readonly body: string;
}

/**
 * Sends one signed S3 request. The default transport talks to real S3 over
 * `fetch`; tests inject a fake to stay credential-free and offline.
 */
export interface S3Transport {
  /** Send `request` and resolve with the S3 response. */
  send(request: S3Request): Promise<S3Response>;
}

/** Explicit AWS credentials. Falls back to `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY`/`AWS_SESSION_TOKEN` if omitted. */
export interface ExperimentalS3IntentGraphStorageCredentials {
  /** AWS access key id. */
  readonly accessKeyId: string;
  /** AWS secret access key. */
  readonly secretAccessKey: string;
  /** Session token for temporary (STS) credentials. */
  readonly sessionToken?: string;
}

/** Options for {@link ExperimentalS3IntentGraphStorage}. */
export interface ExperimentalS3IntentGraphStorageOptions {
  /** S3 bucket to store the graph in. */
  readonly bucket: string;
  /** S3 object key, e.g. `"intent-graph.json"`. */
  readonly key: string;
  /** @default "us-east-1" */
  readonly region?: string;
  /** Explicit credentials; falls back to the standard AWS environment variables. */
  readonly credentials?: ExperimentalS3IntentGraphStorageCredentials;
  /**
   * Custom S3-compatible endpoint, e.g. `"http://localhost:9000"` or
   * `"https://minio.internal:9000"`. Omit for AWS S3 (default).
   */
  readonly endpoint?: string;
  /**
   * Path-style addressing (`https://endpoint/bucket/key`) instead of
   * virtual-hosted-style. Defaults to `true` whenever `endpoint` is set —
   * what MinIO and most self-hosted S3-compatible services require. Pass
   * `false` for a custom endpoint that supports virtual-hosted style (e.g.
   * Cloudflare R2). No effect without `endpoint` — AWS S3 always uses
   * virtual-hosted style.
   */
  readonly forcePathStyle?: boolean;
  /** Override the transport, e.g. to inject a fake for tests. Defaults to a `fetch`-based signed S3 client. */
  readonly transport?: S3Transport;
}

function credentialsFromEnv(): ExperimentalS3IntentGraphStorageCredentials {
  const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
  const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
  if (!accessKeyId || !secretAccessKey) {
    throw new Error(
      "AWS credentials not found: pass `credentials` to ExperimentalS3IntentGraphStorage, or " +
        "set AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY (and AWS_SESSION_TOKEN for temporary credentials).",
    );
  }
  return { accessKeyId, secretAccessKey, sessionToken: process.env.AWS_SESSION_TOKEN };
}

class FetchS3Transport implements S3Transport {
  constructor(
    private readonly region: string,
    private readonly credentials: ExperimentalS3IntentGraphStorageCredentials | undefined,
    private readonly endpoint: string | undefined,
    private readonly forcePathStyle: boolean | undefined,
  ) {}

  async send(request: S3Request): Promise<S3Response> {
    const creds = this.credentials ?? credentialsFromEnv();
    const { scheme, host, path } = resolveS3Endpoint({
      bucket: request.bucket,
      key: request.key,
      region: this.region,
      endpoint: this.endpoint,
      forcePathStyle: this.forcePathStyle,
    });
    const body = request.body ?? "";
    const signed = signS3Request({
      method: request.method,
      host,
      path,
      headers: request.headers,
      body,
      region: this.region,
      accessKeyId: creds.accessKeyId,
      secretAccessKey: creds.secretAccessKey,
      sessionToken: creds.sessionToken,
    });
    const response = await fetch(`${scheme}://${host}${path}`, {
      method: request.method,
      headers: signed.headers,
      body: request.method === "PUT" ? body : undefined,
    });
    const responseBody = await response.text();
    const headers: Record<string, string> = {};
    response.headers.forEach((value, name) => {
      headers[name.toLowerCase()] = value;
    });
    return { status: response.status, headers, body: responseBody };
  }
}

/**
 * S3-backed storage for an {@link IntentGraph}. No SDK dependency — signs
 * requests with a built-in minimal SigV4 implementation (`signS3Request`)
 * and sends them with native `fetch` (ADR-0025). Uses S3 conditional writes
 * (`If-Match`/`If-None-Match` on the object's ETag) for stale-base detection;
 * no bucket versioning required.
 */
export class ExperimentalS3IntentGraphStorage implements ExperimentalIntentGraphStorage {
  private readonly bucket: string;
  private readonly key: string;
  private readonly transport: S3Transport;
  private lastKnownEtag: string | undefined;
  private lastKnownRev: number | undefined;

  constructor(options: ExperimentalS3IntentGraphStorageOptions) {
    this.bucket = options.bucket;
    this.key = options.key;
    this.transport =
      options.transport ??
      new FetchS3Transport(
        options.region ?? "us-east-1",
        options.credentials,
        options.endpoint,
        options.forcePathStyle,
      );
  }

  async load(): Promise<IntentGraph | null> {
    const response = await this.transport.send({
      method: "GET",
      bucket: this.bucket,
      key: this.key,
      headers: {},
    });
    if (response.status === 404) {
      this.lastKnownEtag = undefined;
      this.lastKnownRev = undefined;
      return null;
    }
    if (response.status !== 200) {
      const detail = describeS3Error(response.body);
      throw new Error(
        `S3 GetObject failed for s3://${this.bucket}/${this.key} with status ${response.status}` +
          (detail ? ` (${detail})` : ""),
      );
    }
    const graph = IntentGraph.fromJson(response.body);
    this.lastKnownEtag = response.headers.etag;
    this.lastKnownRev = graph.rev;
    return graph;
  }

  async save(graph: IntentGraph): Promise<void> {
    if (this.lastKnownRev === graph.rev) return;

    const headers: Record<string, string> = this.lastKnownEtag
      ? { "if-match": this.lastKnownEtag }
      : { "if-none-match": "*" };

    const response = await this.transport.send({
      method: "PUT",
      bucket: this.bucket,
      key: this.key,
      headers,
      body: graph.toJson(),
    });

    if (response.status === 412) {
      throw new StaleIntentGraphError(
        `intent graph at s3://${this.bucket}/${this.key} changed since load(); ` +
          `load() again and reapply your changes before saving`,
      );
    }
    if (response.status !== 200) {
      const detail = describeS3Error(response.body);
      throw new Error(
        `S3 PutObject failed for s3://${this.bucket}/${this.key} with status ${response.status}` +
          (detail ? ` (${detail})` : ""),
      );
    }
    this.lastKnownEtag = response.headers.etag;
    this.lastKnownRev = graph.rev;
  }
}

File: src/sdk/python/ratel_ai/intent_graph_storage.py
"""Host-owned persistence for an `IntentGraph` (ADR-0025).

Core stays bytes-in/bytes-out (ADR-0014) — these are thin adapters over
`IntentGraph.to_json()`/`from_json()`/`rev` that live entirely in the SDK.

Both implementations use `rev` for two things: **save-when-changed** (skip the
write if `rev` hasn't moved since the last save) and **stale-base detection**
(raise `StaleIntentGraphError` instead of clobbering a concurrent writer —
single-writer model, detect don't merge).

`ExperimentalS3IntentGraphStorage` has no dependency on `boto3` — it signs
requests with a built-in minimal SigV4 implementation (`ratel_ai._sigv4`) and
sends them with `urllib`.
"""

from __future__ import annotations

import asyncio
import json
import os
import re
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol, runtime_checkable
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from ._native import IntentGraph
from ._sigv4 import resolve_s3_endpoint, sign_s3_request

__all__ = [
    "ExperimentalIntentGraphStorage",
    "ExperimentalLocalFileIntentGraphStorage",
    "ExperimentalS3IntentGraphStorage",
    "ExperimentalS3IntentGraphStorageCredentials",
    "S3Request",
    "S3Response",
    "S3Transport",
    "StaleIntentGraphError",
]


class StaleIntentGraphError(RuntimeError):
    """Another writer saved a newer graph since this storage object's last load()/save()."""


_AWS_ERROR_CODE_RE = re.compile(r"<Code>([^<]*)</Code>")
_AWS_ERROR_MESSAGE_RE = re.compile(r"<Message>([^<]*)</Message>")


def _rev_of(serialized: str) -> int:
    """The ``rev`` carried inside a serialized graph.

    ``save()`` records this rather than re-reading ``graph.rev``, because the
    graph keeps mutating while a save is in flight: ``observe()`` on every
    confirmed invoke, and a centroid rebuild that runs with the GIL released,
    so ``rev`` can move between two native calls. ``to_json()`` serializes
    under one read lock, so the ``rev`` in those bytes is by construction the
    ``rev`` of the content being persisted.
    """
    return int(json.loads(serialized).get("rev", 0))


def _describe_s3_error(body: str) -> str | None:
    """Extract the AWS error code/message from an S3 error response body.

    Standard AWS XML error shape, for a more useful failure message than a bare
    status code — e.g. distinguishing ``SignatureDoesNotMatch`` from
    ``AccessDenied`` on a 403. Returns `None` for a body with no ``<Code>``
    (not an AWS-shaped error).
    """
    code_match = _AWS_ERROR_CODE_RE.search(body)
    if not code_match:
        return None
    code = code_match.group(1)
    message_match = _AWS_ERROR_MESSAGE_RE.search(body)
    return f"{code}: {message_match.group(1)}" if message_match else code


@runtime_checkable
class ExperimentalIntentGraphStorage(Protocol):
    """Host-owned persistence for an `IntentGraph`."""

    async def load(self) -> IntentGraph | None:
        """Load the stored graph, or `None` if nothing has been saved yet."""
        ...

    async def save(self, graph: IntentGraph) -> None:
        """Save `graph`, or skip if unchanged since the last save/load `rev`."""
        ...


class ExperimentalLocalFileIntentGraphStorage:
    """Local JSON file storage for an `IntentGraph` — the default backend.

    Writes atomically (temp file + `os.replace`) so a crash mid-write cannot
    leave a truncated file.
    """

    def __init__(self, path: str | os.PathLike[str]) -> None:
        """Store the graph at `path`. The parent directory must already exist."""
        self._path = Path(path)
        self._last_known_rev: int | None = None

    async def load(self) -> IntentGraph | None:
        """Load the stored graph, or `None` if nothing has been saved yet."""
        try:
            text = await asyncio.to_thread(self._path.read_text)
        except FileNotFoundError:
            self._last_known_rev = None
            return None
        graph = IntentGraph.from_json(text)
        self._last_known_rev = graph.rev
        return graph

    async def save(self, graph: IntentGraph) -> None:
        """Save `graph`, or skip if unchanged since the last save/load `rev`."""
        if self._last_known_rev == graph.rev:
            return

        body = graph.to_json()
        rev = _rev_of(body)

        disk_rev = await asyncio.to_thread(self._read_disk_rev)
        if disk_rev != self._last_known_rev:
            expected = self._last_known_rev if self._last_known_rev is not None else "none"
            found = disk_rev if disk_rev is not None else "none"
            raise StaleIntentGraphError(
                f"intent graph at {self._path} changed since load() (on-disk rev {found}, "
                f"expected {expected}); load() again and reapply your changes before saving"
            )

        await asyncio.to_thread(self._write_atomic, body)
        self._last_known_rev = rev

    def _read_disk_rev(self) -> int | None:
        try:
            text = self._path.read_text()
        except FileNotFoundError:
            return None
        rev = json.loads(text).get("rev", 0)
        return int(rev)

    def _write_atomic(self, contents: str) -> None:
        fd, tmp_path = tempfile.mkstemp(dir=self._path.parent, prefix=".tmp-")
        try:
            with os.fdopen(fd, "w") as f:
                f.write(contents)
            os.replace(tmp_path, self._path)
        except BaseException:
            if os.path.exists(tmp_path):
                os.remove(tmp_path)
            raise


@dataclass(frozen=True)
class S3Request:
    """One S3 REST request, as `S3Transport` sends it."""

    method: str  # "GET" | "PUT"
    bucket: str
    key: str
    headers: dict[str, str]
    body: str | None = None


@dataclass(frozen=True)
class S3Response:
    """Response from an `S3Transport` call."""

    status: int
    headers: dict[str, str]
    body: str


@runtime_checkable
class S3Transport(Protocol):
    """Sends one signed S3 request.

    The default transport talks to real S3 over `urllib`; tests inject a fake
    to stay credential-free and offline.
    """

    async def send(self, request: S3Request) -> S3Response:
        """Send one S3 request and return its response."""
        ...


@dataclass(frozen=True)
class ExperimentalS3IntentGraphStorageCredentials:
    """Explicit AWS credentials.

    Falls back to ``AWS_ACCESS_KEY_ID``/``AWS_SECRET_ACCESS_KEY``/``AWS_SESSION_TOKEN``
    if omitted.
    """

    access_key_id: str
    secret_access_key: str
    session_token: str | None = None


def _credentials_from_env() -> ExperimentalS3IntentGraphStorageCredentials:
    access_key_id = os.environ.get("AWS_ACCESS_KEY_ID")
    secret_access_key = os.environ.get("AWS_SECRET_ACCESS_KEY")
    if not access_key_id or not secret_access_key:
        raise RuntimeError(
            "AWS credentials not found: pass `credentials` to "
            "ExperimentalS3IntentGraphStorage, or set AWS_ACCESS_KEY_ID / "
            "AWS_SECRET_ACCESS_KEY (and AWS_SESSION_TOKEN for temporary credentials)."
        )
    return ExperimentalS3IntentGraphStorageCredentials(
        access_key_id=access_key_id,
        secret_access_key=secret_access_key,
        session_token=os.environ.get("AWS_SESSION_TOKEN"),
    )


class _UrllibS3Transport:
    def __init__(
        self,
        region: str,
        credentials: ExperimentalS3IntentGraphStorageCredentials | None,
        endpoint: str | None = None,
        force_path_style: bool | None = None,
    ) -> None:
        self._region = region
        self._credentials = credentials
        self._endpoint = endpoint
        self._force_path_style = force_path_style

    async def send(self, request: S3Request) -> S3Response:
        return await asyncio.to_thread(self._send_sync, request)

    def _send_sync(self, request: S3Request) -> S3Response:
        creds = self._credentials or _credentials_from_env()
        target = resolve_s3_endpoint(
            bucket=request.bucket,
            key=request.key,
            region=self._region,
            endpoint=self._endpoint,
            force_path_style=self._force_path_style,
        )
        body = request.body or ""
        signed = sign_s3_request(
            method=request.method,
            host=target.host,
            path=target.path,
            headers=dict(request.headers),
            body=body,
            region=self._region,
            access_key_id=creds.access_key_id,
            secret_access_key=creds.secret_access_key,
            session_token=creds.session_token,
        )
        url = f"{target.scheme}://{target.host}{target.path}"
        data = body.encode("utf-8") if request.method == "PUT" else None
        http_request = Request(url, data=data, headers=signed.headers, method=request.method)
        try:
            with urlopen(http_request) as response:  # noqa: S310 - S3 REST API, https only
                response_body = response.read().decode("utf-8")
                headers = {k.lower(): v for k, v in response.headers.items()}
                return S3Response(status=response.status, headers=headers, body=response_body)
        except HTTPError as error:
            response_body = error.read().decode("utf-8")
            headers = {k.lower(): v for k, v in error.headers.items()} if error.headers else {}
            return S3Response(status=error.code, headers=headers, body=response_body)


class ExperimentalS3IntentGraphStorage:
    """S3-backed storage for an `IntentGraph`.

    No SDK dependency — signs requests with a built-in minimal SigV4
    implementation (`ratel_ai._sigv4.sign_s3_request`) and sends them with
    `urllib` (ADR-0025). Uses S3 conditional writes (``If-Match``/
    ``If-None-Match`` on the object's ETag) for stale-base detection; no
    bucket versioning required.
    """

    def __init__(
        self,
        *,
        bucket: str,
        key: str,
        region: str = "us-east-1",
        credentials: ExperimentalS3IntentGraphStorageCredentials | None = None,
        endpoint: str | None = None,
        force_path_style: bool | None = None,
        transport: S3Transport | None = None,
    ) -> None:
        """Store the graph at `s3://{bucket}/{key}` in `region`.

        `credentials` defaults to `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY`/
        `AWS_SESSION_TOKEN` if omitted. `endpoint` points at a custom
        S3-compatible service (e.g. `"http://localhost:9000"` for MinIO);
        omit for AWS S3. `force_path_style` defaults to `True` whenever
        `endpoint` is set (what MinIO and most self-hosted services
        require) — pass `False` for a custom endpoint that supports
        virtual-hosted style. `transport` overrides how requests are sent —
        inject a fake for tests.
        """
        self._bucket = bucket
        self._key = key
        self._transport: S3Transport = transport or _UrllibS3Transport(
            region, credentials, endpoint, force_path_style
        )
        self._last_known_etag: str | None = None
        self._last_known_rev: int | None = None

    async def load(self) -> IntentGraph | None:
        """Load the stored graph, or `None` if the object does not exist."""
        response = await self._transport.send(
            S3Request(method="GET", bucket=self._bucket, key=self._key, headers={})
        )
        if response.status == 404:
            self._last_known_etag = None
            self._last_known_rev = None
            return None
        if response.status != 200:
            detail = _describe_s3_error(response.body)
            raise RuntimeError(
                f"S3 GetObject failed for s3://{self._bucket}/{self._key} "
                f"with status {response.status}" + (f" ({detail})" if detail else "")
            )
        graph = IntentGraph.from_json(response.body)
        self._last_known_etag = response.headers.get("etag")
        self._last_known_rev = graph.rev
        return graph

    async def save(self, graph: IntentGraph) -> None:
        """Save `graph`, or skip if unchanged since the last save/load `rev`."""
        if self._last_known_rev == graph.rev:
            return

        headers = (
            {"if-match": self._last_known_etag}
            if self._last_known_etag is not None
            else {"if-none-match": "*"}
        )
        body = graph.to_json()
        rev = _rev_of(body)

        response = await self._transport.send(
            S3Request(
                method="PUT",
                bucket=self._bucket,
                key=self._key,
                headers=headers,
                body=body,
            )
        )
        if response.status == 412:
            raise StaleIntentGraphError(
                f"intent graph at s3://{self._bucket}/{self._key} changed since load(); "
                "load() again and reapply your changes before saving"
            )
        if response.status != 200:
            detail = _describe_s3_error(response.body)
            raise RuntimeError(
                f"S3 PutObject failed for s3://{self._bucket}/{self._key} "
                f"with status {response.status}" + (f" ({detail})" if detail else "")
            )
        self._last_known_etag = response.headers.get("etag")
        self._last_known_rev = rev
